### DISCUSSION TITLE
Use alternative cloud providers with maestral

### DISCUSSION THREAD

**Discussion by hlibt** [samschott/maestral#400] [Ideas]
I am wondering if it could be possible to decouple syncronization backend from the rest of maestral and use  [rclone](https://rclone.org/) to implement sync. I personally would love to have a maestral like app for google drive on linux.

**Comment by samschott** [comment]
Uh, that is a good idea but also quite a large expansion of this project. Looks like it would end up being a FOSS version of [insync](https://www.insynchq.com). I'm unlikely to be able to tackle this on my own.

As it stands, all code that interfaces with Dropbox APIs directly lives in the `client` module. As long as this `client` functionality can be implemented with the APIs of other cloud providers, the basic layer of abstraction to separate the sync engine from the the actual client API already exists.

For sync functionality, the Dropbox API is very convenient because it allows the client to wait for or query for changes to a folder since a specific "cursor". From a quick glance at the [docs](https://developers.google.com/drive/api/v3/about-changes), it appears that the Google Drive API provides something similar. AFAIK, `rsync` does not abstract this functionality because it is not targeting background syncing. So we will likely end up writing clients for each cloud provider (not too much work, the sync engine is the more complex part). But I've only just looked at rsync for the first time, happy to be proven wrong...

Apart from that, the GUI would need to be updated. But this is mostly just tedious instead of difficult.

**Reply by hlibt** [reply]
Just to make sure we are talking about the same, I was referring to https://rclone.org/ .

**Reply by samschott** [reply]
Sorry, yes, rclone, not rsync. But the same still holds, I don't they have proper support to wait for remote changes (except for polling). An initial discussion about this has been started at https://github.com/rclone/rclone/issues/4381.

I am also not sure how `rclone sync` is implemented when syncing `remote -> local`. Does it fetch a list of changes since the last sync command? Or does it fetch all remote entries and then compare them to the local ones?
